File: QueueWithStack.java
// Time Complexity : Push - O(1), Pop - O(1) (amortized), Peek - O(1) (amortized), empty - O(1)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

// logic: push the element into the "in" stack when pop is called for the first time
// push all elements from "in" to "out" stack. Pop elements from "out" stack and push elements into "in" stack.
// Use "out" stack for peek() and check if "in" and "out" stacks are empty or not for the empty().

class QueueWithStack {

    Stack<Integer> in;
    Stack <Integer> out;

    public QueueWithStack() {
        in = new Stack<>();
        out = new Stack<>();
    }

    public void push(int x) {
        in.push(x);

    }

    public int pop() {
        peek();
        return out.pop();
    }

    public int peek() {
        if(out.isEmpty()){
            while(!in.isEmpty()){
                out.push(in.pop());
            }
        }
        return out.peek();
    }

    public boolean empty() {
        return in.isEmpty() && out.isEmpty();
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */

File: queue_with_stack.js
// Time Complexity : Push, Pop, Peek O(1)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

var MyQueue = function () {
  this.in_stack = [];
  this.out_stack = [];
};

/**
 * @param {number} x
 * @return {void}
 */
MyQueue.prototype.push = function (x) {
  this.in_stack.push(x);
};

/**
 * @return {number}
 */
MyQueue.prototype.pop = function () {
  if (this.out_stack.length == 0) {
    this.transfer();
  }
  return this.out_stack.pop();
};

/**
 * @return {number}
 */
MyQueue.prototype.peek = function () {
  if (this.out_stack.length == 0) {
    this.transfer();
  }
  return this.out_stack[this.out_stack.length - 1];
};

MyQueue.prototype.transfer = function () {
  while (this.in_stack.length != 0) {
    this.out_stack.push(this.in_stack.pop());
  }
};

/**
 * @return {boolean}
 */
MyQueue.prototype.empty = function () {
  return this.in_stack.length == 0 && this.out_stack.length == 0;
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * var obj = new MyQueue()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */

test("Scenario #1:", () => {
  var obj = new MyQueue();
  obj.push(1);
  obj.push(2);
  obj.push(3);
  obj.push(4);

  expect(obj.pop()).toStrictEqual(1);
  expect(obj.peek()).toStrictEqual(2);
  expect(obj.empty()).toStrictEqual(false);
});
